analyze-ms-stats/calc-manuscript-stats.py

```diff
@@ -0,0 +1,126 @@
+import json
+import pandas as pd
+import matplotlib
+import argparse
+import time
+import os.path
+import multiprocessing
+import subprocess
+
+def analyze_commit(commit):
+    """Access files and data in variables.json associated with each commit
+    Accepts commit ID as string
+    Returns list of 5 statistics"""
+
+    variablesCommand = "git show " + commit + ":./variables.json"
+    variables = json.loads(subprocess.getoutput(variablesCommand))
+
+    date = variables['pandoc']['date-meta']
+    clean_date = variables['manubot']['date']
+    num_authors = len(variables['manubot']['authors'])
+    word_count = variables['manubot']['manuscript_stats']['word_count']
+
+    # Access files and data in references.json associated with each commit
+    referencesCommand = "git show " + commit + ":./references.json"
+    references = json.loads(subprocess.getoutput(referencesCommand))
+    num_ref = len(references)
+
+    return([date, clean_date, num_authors, word_count, num_ref])
+
+def main(args):
+    '''Extract statistics from the output branch log'''
+
+    start_time = time.time()
+    print(f"Using {multiprocessing.cpu_count()} CPUs")
+
+    # Read in list of all commits on this branch
+    with open(args.commit_list, "r") as commitFile:
+        commits = [c.strip() for c in commitFile.read().splitlines()]
+
+    # If this analysis has been run before, load in the list of commits analyzed
+    # and only analyze new commits
+    # Assumes no commits will be added retrospectively (to take advantage of linearity)
+    priorData = None
+    if os.path.exists(args.output_table):
+        priorData = pd.read_csv(args.output_table)
+        oldCommits = priorData["commit"].tolist()
+        priorData = priorData.set_index("commit")
+
+        if len(commits) > len(oldCommits):
+            start_old = commits.index(oldCommits[0])
+            commits = commits[:start_old]
+            print(f"{len(commits)} new commits")
+        else:
+            exit("No new commits")
+
+    # Access the variables.json and references.json files associated with each commit and store in dictionary
+    pool = multiprocessing.Pool(processes=multiprocessing.cpu_count())
```

Comment:
I haven't used pools in a long time.  I see some examples using `with` statements (https://stackoverflow.com/questions/45718546/with-clause-for-multiprocessing-in-python) but do not know whether that is necessary.  Perhaps it is more robust in case the commit processing has an unexpected error.

Reply:
Yes, I think that's the only advantage of using `with` (not sure what happens if any of the tasks fails and you didn't close the pool). Usually, I prefer to use `futures`, like in [this function](https://github.com/hakyimlab/phenomexcan/blob/38390ac21987f1e72835c42919c53abd1a35cb7e/scripts/000_spredixcan/utils/generate_jobs.py#L78), `write_phenotypes_jobs`. But if this is already working fine, no need to change.